src/Application.php

```diff
@@ -52,19 +53,39 @@ public function run()
         };
 
         // Trigger route event
+        $event->setName(MvcEvent::EVENT_ROUTE);
         try {
-            $result = $events->trigger(MvcEvent::EVENT_ROUTE, $event, $shortCircuit);
+            $result = $events->triggerEventUntil($shortCircuit, $event);
+        } catch (Throwable $e) {
+            $event->setName(MvcEvent::EVENT_DISPATCH_ERROR);
+            $event->setError(self::ERROR_EXCEPTION);
+            $event->setParam('exception', $e);
+            $result = $events->triggerEvent($event);
+
+            $response = $result->last();
+            if ($response instanceof ResponseInterface) {
+                $event->setName(MvcEvent::EVENT_FINISH);
+                $event->setTarget($this);
+                $event->setResponse($response);
+                $this->response = $response;
+                $events->triggerEvent($event);
+                return $this;
+            }
+
+            return $this->completeRequest($event);
```

Comment:
can we extract this part to separate method to avoid code duplication? is it exactly the same in `catch (Exception $e)` ?